File: backend/src/services/payfastService.js
const crypto = require('crypto');
const axios = require('axios');
const dns = require('dns').promises;
const admin = require('../../firebase');

/**
 * PayFast Integration Service
 * Based on official PayFast documentation: https://developers.payfast.co.za/docs
 */
class PayfastService {
  constructor() {
    require('dotenv').config({ path: '.env' });

    let firebaseConfig = {};
    try {
      const functions = require('firebase-functions');
      firebaseConfig = functions.config().payfast || {};
    } catch (err) {
    }

    this.config = {
      merchantId: String(firebaseConfig.merchant_id || process.env.PAYFAST_MERCHANT_ID || '').trim(),
      merchantKey: String(firebaseConfig.merchant_key || process.env.PAYFAST_MERCHANT_KEY || '').trim(),
      passphrase: '', 
      returnUrl: String(firebaseConfig.return_url || process.env.PAYFAST_RETURN_URL || '').trim(),
      cancelUrl: String(firebaseConfig.cancel_url || process.env.PAYFAST_CANCEL_URL || '').trim(),
      notifyUrl: String(firebaseConfig.notify_url || process.env.PAYFAST_NOTIFY_URL || '').trim(),
      testMode: String(firebaseConfig.test_mode || process.env.PAYFAST_TEST_MODE) === 'true'
    };

    this.paymentUrl = this.config.testMode
      ? 'https://sandbox.payfast.co.za/eng/process'
      : 'https://www.payfast.co.za/eng/process';
  }

  /**
   * Generates the PayFast signature following their exact documentation
   * https://developers.payfast.co.za/docs#signature-generation
   */
  generateSignature(data, passPhrase = null) {
    // Sort keys alphabetically (IMPORTANT)
    const sortedKeys = Object.keys(data).sort();

    // Build parameter string
    let pfOutput = "";
      for (let key of sortedKeys) {
        if(data.hasOwnProperty(key)){
          if (data[key] !== "") {
            pfOutput +=`${key}=${encodeURIComponent(data[key].trim()).replace(/%20/g, "+")}&`
          }
        }
      }

    // Remove last ampersand
    let getString = pfOutput.slice(0, -1);

    // Generate MD5 hash in lowercase
    const signature = crypto.createHash("md5").update(getString).digest("hex");

    console.log('\n=== PayFast Signature Generation ===');
    console.log('Parameter string:', getString);
    console.log('Passphrase used: NO');
    console.log('Generated signature:', signature);
    console.log('===================================\n');

    return signature;
  }

  /**
   * Creates the complete payment request object to be sent to the frontend.
   */
  createPaymentRequest(orderData, orderId, userId) {
    try {
      if (!this.config.merchantId || !this.config.merchantKey) {
        throw new Error('PayFast merchant_id or merchant_key is not configured.');
      }

      const fullName = String(orderData.deliveryAddress?.name || 'Customer').trim();
      const nameParts = fullName.split(/\s+/);
      const firstName = (nameParts[0] || 'Customer').substring(0, 100);
      const lastName = (nameParts.slice(1).join(' ') || 'User').substring(0, 100);
      const email = (orderData.email || `${userId}@locals-za.co.za`).substring(0, 255);
      const amountString = parseFloat(orderData.total || 0).toFixed(2);

      if (parseFloat(amountString) <= 0) {
        throw new Error('Invalid order amount. Must be greater than zero.');
      }

      // Step 1: Build data FOR SIGNATURE (WITH merchant_key, matching PayFast order)
      const dataForSignature = {
        merchant_id: this.config.merchantId,
        merchant_key: this.config.merchantKey,
        return_url: `${this.config.returnUrl}/${orderId}`,
        cancel_url: `${this.config.cancelUrl}/${orderId}`,
        notify_url: this.config.notifyUrl,
        name_first: firstName,
        name_last: lastName,
        email_address: email,
        m_payment_id: orderId,
        amount: amountString,
        item_name: `LocalsZA Order #${orderId.slice(-8)}`.substring(0, 255)
      };

      // Step 2: Generate signature
      const signature = this.generateSignature(dataForSignature, this.config.passphrase);

      // Step 3: Build final form data (SAME ORDER as dataForSignature, add signature at end)
      const formData = {
        merchant_id: this.config.merchantId,
        merchant_key: this.config.merchantKey,
        return_url: `${this.config.returnUrl}/${orderId}`,
        cancel_url: `${this.config.cancelUrl}/${orderId}`,
        notify_url: this.config.notifyUrl,
        name_first: firstName,
        name_last: lastName,
        email_address: email,
        m_payment_id: orderId,
        amount: amountString,
        item_name: `LocalsZA Order #${orderId.slice(-8)}`.substring(0, 255),
        signature: signature
      };

      console.log('=== PayFast Payment Request Summary ===');
      console.log('Data for signature (WITH merchant_key):', dataForSignature);
      console.log('Form data to PayFast (WITH merchant_key + signature):', formData);
      console.log('Passphrase:NOT USED');
      console.log('========================================');

      return {
        formData: formData,
        url: this.paymentUrl,
        paymentId: orderId
      };

    } catch (error) {
      console.error('Error creating PayFast payment request:', error);
      throw error;
    }
  }

  /**
   * Verify ITN signature
   */
  verifySignature(data, receivedSignature) {
    try {
      // Remove signature from data for recalculation
      const dataForVerification = { ...data };
      delete dataForVerification.signature;

      // Sort keys alphabetically
      const sortedKeys = Object.keys(dataForVerification).sort();

      // Build parameter string (same as payment request)
      let paramString = "";
      for (let key of sortedKeys) {
        if (dataForVerification.hasOwnProperty(key)) {
          if (dataForVerification[key] !== "") {
            paramString += `${key}=${encodeURIComponent(String(dataForVerification[key]).trim()).replace(/%20/g, "+")}&`;
          }
        }
      }

      // Remove last ampersand
      paramString = paramString.slice(0, -1);

      // Generate MD5 hash
      const calculatedSignature = crypto.createHash("md5").update(paramString).digest("hex");

      console.log('=== PayFast Signature Verification ===');
      console.log('Received:', receivedSignature);
      console.log('Calculated:', calculatedSignature);
      console.log('Match:', calculatedSignature === receivedSignature);
      console.log('=====================================');

      return calculatedSignature === receivedSignature;
    } catch (error) {
      console.error('Signature verification error:', error);
      return false;
    }
  }

  /**
   * Parse ITN data from POST body
   */
  parseItnData(body) {
    const data = {};
    const bodyString = body.toString();
    
    bodyString.split('&').forEach(pair => {
      const [key, value] = pair.split('=');
      if (key && value !== undefined) {
        data[decodeURIComponent(key)] = decodeURIComponent(value.replace(/\+/g, ' '));
      }
    });

    return data;
  }

  /**
   * Validate request IP is from PayFast
   */
  async validateIp(req) {
    try {
      const validHosts = [
        'www.payfast.co.za',
        'sandbox.payfast.co.za',
        'w1w.payfast.co.za',
        'w2w.payfast.co.za'
      ];

      let validIps = [];
      const pfIp = req.headers['x-forwarded-for'] || req.connection.remoteAddress;

      console.log('Validating IP:', pfIp);

      // Lookup all PayFast IPs
      for (const host of validHosts) {
        try {
          const addresses = await dns.lookup(host, { all: true });
          const ips = addresses.map(item => item.address);
          validIps = [...validIps, ...ips];
          console.log(`IPs for ${host}:`, ips);
        } catch (err) {
          console.error(`DNS lookup failed for ${host}:`, err.message);
        }
      }

      // Remove duplicates
      const uniqueIps = [...new Set(validIps)];
      console.log('Valid PayFast IPs:', uniqueIps);

      const isValid = uniqueIps.includes(pfIp);
      console.log(`IP ${pfIp} is ${isValid ? 'VALID' : 'INVALID'}`);

      return isValid;
    } catch (error) {
      console.error('IP validation error:', error);
      return false;
    }
  }

  /**
   * Validate payment with PayFast server
   */
  async validateWithPayfast(data) {
    try {
      const validateUrl = this.config.testMode
        ? 'https://sandbox.payfast.co.za/eng/query/validate'
        : 'https://www.payfast.co.za/eng/query/validate';

      // Build form data for PayFast validation
      const params = Object.keys(data)
        .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(String(data[key]))}`)
        .join('&');

      console.log('Sending validation to PayFast...');

      const response = await axios.post(validateUrl, params, {
        headers: { 
          'Content-Type': 'application/x-www-form-urlencoded',
          'User-Agent': 'LocalsZA-PaymentProcessor/1.0'
        },
        timeout: 10000
      });

      const isValid = response.data.trim() === 'VALID';
      console.log('PayFast validation response:', response.data.trim());

      return isValid;
    } catch (error) {
      console.error('PayFast validation error:', error.message);
      return false;
    }
  }

  /**
   * Process ITN callback with full security validation
   */
  async processItn(data, req) {
    try {
      console.log('=== Processing PayFast ITN ===');
      console.log('Payment ID:', data.m_payment_id);
      console.log('Status:', data.payment_status);

      // Step 1: Validate IP address
      const validIp = await this.validateIp(req);
      if (!validIp) {
        console.error('Invalid IP address for ITN');
        return { success: false, error: 'Invalid IP address' };
      }

      // Step 2: Verify signature
      if (!this.verifySignature(data, data.signature)) {
        console.error('Invalid signature');
        return { success: false, error: 'Invalid signature' };
      }

      // Step 3: Validate with PayFast server
      const isValid = await this.validateWithPayfast(data);
      if (!isValid) {
        console.error('Server validation failed');
        return { success: false, error: 'Server validation failed' };
      }

      console.log('============================');
      console.log('All validations passed');

      // Process the payment...
      const orderId = data.m_payment_id;
      if (!orderId) {
        return { success: false, error: 'Missing order ID' };
      }

      // Update order in database
      const orderRef = admin.firestore().collection('orders').doc(orderId);
      const orderDoc = await orderRef.get();

      if (!orderDoc.exists) {
        return { success: false, error: 'Order not found' };
      }

      // Log ITN for audit
      await admin.firestore().collection('payment_notifications').add({
        orderId,
        paymentData: data,
        receivedAt: admin.firestore.FieldValue.serverTimestamp(),
        environment: this.config.testMode ? 'sandbox' : 'production',
        verified: true,
        ipAddress: req.headers['x-forwarded-for'] || req.connection.remoteAddress
      });

      // Update order status
      const paymentStatus = data.payment_status;
      const updateData = {
        paymentData: data,
        paymentUpdatedAt: admin.firestore.FieldValue.serverTimestamp(),
        paymentVerified: true
      };

      if (paymentStatus === 'COMPLETE') {
        updateData.status = 'pending';
        updateData.paymentStatus = 'paid';
        updateData.paymentCompleted = true;
        updateData.paymentCompletedAt = admin.firestore.FieldValue.serverTimestamp();
        updateData.pf_payment_id = data.pf_payment_id;
      } else if (paymentStatus === 'FAILED') {
        updateData.status = 'payment_failed';
        updateData.paymentStatus = 'failed';
      } else if (paymentStatus === 'CANCELLED') {
        updateData.status = 'cancelled';
        updateData.paymentStatus = 'cancelled';
      }

      await orderRef.update(updateData);

      return {
        success: true,
        orderId,
        status: paymentStatus,
        userId: orderDoc.data().userId
      };

    } catch (error) {
      console.error('PayFast ITN processing error:', error);
      return { success: false, error: error.message };
    }
  }
}

module.exports = new PayfastService();

File: api/Services/PayfastService.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using LocalsZaApi.Models;

namespace LocalsZaApi.Services;

public class PayfastService
{
    private readonly IConfiguration _config;
    private readonly ILogger<PayfastService> _log;

    public PayfastService(IConfiguration config, ILogger<PayfastService> log)
    {
        _config = config;
        _log = log;
    }

    private string MerchantId  => _config["Payfast:MerchantId"]  ?? throw new InvalidOperationException("Payfast:MerchantId missing");
    private string MerchantKey => _config["Payfast:MerchantKey"] ?? throw new InvalidOperationException("Payfast:MerchantKey missing");
    private string Passphrase  => _config["Payfast:Passphrase"]  ?? "";
    private bool   IsTest      => bool.Parse(_config["Payfast:IsTest"] ?? "false");
    private string PaymentUrl  => IsTest
        ? "https://sandbox.payfast.co.za/eng/process"
        : "https://www.payfast.co.za/eng/process";

    /// <summary>
    /// Build the payload + URL the frontend should POST to (PayFast consumer checkout).
    /// </summary>
    public PayfastInitiateResult Initiate(Order order, string? customerEmail, string baseAppUrl)
    {
        var amount = order.Total.ToString("F2", CultureInfo.InvariantCulture);

        if (decimal.Parse(amount, CultureInfo.InvariantCulture) <= 0)
            throw new InvalidOperationException("Order amount must be greater than zero");

        var fullName  = (order.CustomerName?.Trim() is { Length: > 0 } n) ? n : "Customer";
        var parts     = fullName.Split(' ', StringSplitOptions.RemoveEmptyEntries);
        var firstName = parts[0][..Math.Min(100, parts[0].Length)];
        var lastName  = parts.Length > 1
            ? string.Join(" ", parts.Skip(1))[..Math.Min(100, string.Join(" ", parts.Skip(1)).Length)]
            : "User";

        var email    = (customerEmail is { Length: > 0 } e) ? e[..Math.Min(255, e.Length)] : $"guest@locals-za.co.za";
        var tail     = order.Id.Length <= 8 ? order.Id : order.Id[^8..];
        var itemName = $"LocalsZA Order #{tail}";

        // Notify URL points at the API directly, not the SPA base URL
        var notifyUrl = _config["Payfast:NotifyUrl"]
            ?? throw new InvalidOperationException("Payfast:NotifyUrl missing");

        var dataForSignature = new Dictionary<string, string>
        {
            ["merchant_id"]    = MerchantId,
            ["merchant_key"]   = MerchantKey,
            ["return_url"]     = $"{baseAppUrl}/order/payment/success/{order.Id}",
            ["cancel_url"]     = $"{baseAppUrl}/order/payment/cancelled/{order.Id}",
            ["notify_url"]     = notifyUrl,
            ["name_first"]     = firstName,
            ["name_last"]      = lastName,
            ["email_address"]  = email,
            ["m_payment_id"]   = order.Id,
            ["amount"]         = amount,
            ["item_name"]      = itemName,
        };

        var signature = GenerateSignature(dataForSignature, Passphrase);

        var fields = new Dictionary<string, string>(dataForSignature)
        {
            ["signature"] = signature
        };

        _log.LogInformation("PayFast initiate: orderId={OrderId} amount={Amount} isTest={IsTest}",
            order.Id, amount, IsTest);

        return new PayfastInitiateResult(PostUrl: PaymentUrl, Fields: fields, Signature: signature);
    }

    /// <summary>
    /// Verify the signature on a PayFast ITN (Instant Transaction Notification).
    /// </summary>
    public bool VerifyItn(IReadOnlyDictionary<string, string> form)
    {
        if (!form.TryGetValue("signature", out var received))
        {
            _log.LogWarning("PayFast ITN: no signature field present");
            return false;
        }

        var data = form
            .Where(kv => kv.Key != "signature")
            .ToDictionary(kv => kv.Key, kv => kv.Value);

        var computed = GenerateSignature(data, Passphrase);
        var match = string.Equals(computed, received, StringComparison.OrdinalIgnoreCase);

        if (!match)
            _log.LogWarning("PayFast ITN signature mismatch. received={Received} computed={Computed}",
                received, computed);

        return match;
    }

    // ---------- signature ----------

    /// <summary>
    /// PayFast MD5 signature: sort keys alphabetically, URL-encode values,
    /// join with &amp;, optionally append &amp;passphrase=..., then MD5 lowercase hex.
    /// https://developers.payfast.co.za/docs#signature-generation
    /// </summary>
    private static string GenerateSignature(IDictionary<string, string> data, string? passphrase)
    {
        var parts = data
            .OrderBy(kv => kv.Key, StringComparer.Ordinal)
            .Where(kv => !string.IsNullOrEmpty(kv.Value))
            .Select(kv => $"{kv.Key}={Uri.EscapeDataString(kv.Value.Trim()).Replace("%20", "+")}");

        var pfOutput = string.Join("&", parts);

        if (!string.IsNullOrEmpty(passphrase))
        {
            var encodedPassphrase = Uri.EscapeDataString(passphrase.Trim()).Replace("%20", "+");
            pfOutput += $"&passphrase={encodedPassphrase}";
        }

        var hash = MD5.HashData(Encoding.UTF8.GetBytes(pfOutput));
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}

public record PayfastInitiateResult(string PostUrl, IDictionary<string, string> Fields, string Signature);
